README.md
# UnderlyingEvent

This README will decribe the details of how to run and create the jet, skim and ue trees. 


RooUnfold/History
History of RooUnfold     $Id: History 345 2013-07-29 23:38:49Z T.J.Adye $
====================

Tim Adye         29-Jul-2013
  o Update local copy of TSVDUnfold from ROOT 5.34.00, which fixes some memory
    leaks. (Actually used ROOT 5.34.06, which has a comment spelling fix.)
  o Use weighted errors in SVD systematic error if response matrix has error
    weights set (Sumw2() called on input response matrix histogram). This uses
    a private version of TSVDUnfold from Kerstin Tackmann, responding to a
    report by Brian Page.
  o Preserve Sumw2() setting in histograms created by RooUnfoldResponse::H2H1D,
    RooUnfoldResponse::HresponseNoOverflow, and RooUnfold::HistNoOverflow.
  o Don't use TUnfoldSys for systematic errors with TUnfold if not available
    (ROOT 5.22 - previously didn't have TUnfold, later versions had TUnfoldSys).

Tim Adye          2-Jul-2013
  o RooUnfoldTest seed=0 uses a unique seed for each job.

Tim Adye          8-Jan-2013
  o Implement RooUnfoldResponse::Merge so hadd and TFileMerger can merge
    response matrix objects.

Tim Adye          9-Jul-2012
  o Fix rootmap creation with GCC 4.6. [Bug reported by Gavin Hesketh.]
  o Fix double delete in ROOT 5.34 TSVDUnfold.

Tim Adye          6-Jul-2012
  o Change RooUnfold::PrintTable default to show errors as returned by the
    last unfolding (or else kErrors). If kNoError specified, don't show
    error at all. Previously sqrt(N) was shown for the error unless specified
    otherwise in the withError parameter. [Bug reported by Stefan Kluth.]

Tim Adye         12-Jun-2012
  o Add RooUnfoldResponse::MakeFoldingFunction which wraps a TF1 to apply the
    response matrix to a user parametric function. This can be fitted to the
    measured distribution as an alternative to unfolding.
  o RooUnfoldTest fit=1 tests this (test function definitions only implemented
    for the 1D case).
  o Fix some specific compiler warnings.

Tim Adye          8-Jun-2012
  o Fix some GCC 4.6 warnings.

Tim Adye          1-Jun-2012
  o GNUMakefile sets LDFLAGS, even if Makefile.arch not found. Change suggested
    by Martin Woudstra.
  o Add .rootrc file to define ACLiC.IncludePaths, so user doesn't have to
    say ".include src" to run code with ACLiC.

Tim Adye         22-Mar-2012
  o RooUnfoldResponse::ReplaceAxis: keep axis name when copying from another
    axis. Unfolded histogram was getting the x-axis called 'yaxis' (from truth
    axis of response matrix), which breaks TAxis::SetRangeUser.
    Bug reported by Petr Balek.

Tim Adye         27-Jan-2012
  o RooUnfoldSvd deals with fakes by subtracting from the measured input, rather
    than adding an extra effectless-causes bin, which didn't work very well if
    it was much different from its neighbouring bin.
    Giovanni Marchiori reported this problem.
  o Fix PrintTable "total pull". Bug reported by Giovanni Marchiori.
  o Level 2 verbose printout shows results table with algorithm binning.
    Uses new RooUnfold::PrintTable methods.
  o RooUnfoldResponse::Print shows response matrix contents.
  o Matrix debug printout uses new RooUnfoldResponse::PrintMatrix method
    (borrowed from RooUnfoldTestHarness), which is neater than TMatrix::Print.

Tim Adye         15-Dec-2011
  o RooUnfoldResponse::FindBin is now public to help users fill their own
    response matrix histograms. For consistency, there's also now a trivial 1D
    version.
  o 2+3D RooUnfoldResponse::Fill no longer assumes _res has the expected axis
    ranges, since this may be a user's histogram.
  o RooUnfoldTest "seed" parameter allows random number seed to be set.

Tim Adye         17-Nov-2011
  o Fix for ROOT 5.12 and before, which doesn't have TH1::GetEffectiveEntries().
  o Remove std::vector definitions from RooUnfold dictionary.
    These gave warnings when the library was loaded into ROOT, where they were
    already defined.
  o GNUMakefile runs RooUnfold_LinkDef.h through the preprocessor before running
    rlibmap, so deselected classes aren't included in the rootmap.
  o Fix GNUMakefile for gmake 3.80 and before (eg. in SL4) which doesn't allow
    "else ifneq".
  o Remove spurious "cd" in rootcint make rule.
  o Remove gSystem->Load("libRooUnfold") from examples. That should happen
    automatically now that we have a libRooUnfold.rootmap.
  o Tidy up and expand help at the top of the GNUmakefile.
  o GNUmakefile can now build user programs in the current directory.

Tim Adye         15-Nov-2011
  o Protect against null RooUnfoldResponse fakes histogram pointer. This can
    arise if reading an object created with an old version of RooUnfold
    (prior to 1.1.0).

Tim Adye         14-Nov-2011
  o Fix trivial compiler warning.

Tim Adye          4-Nov-2011
  o GNUMakefile update for ROOT 5.32, which moved Makefile.arch from the test
    to the etc directory and renamed $(ROOTCONFIG) to $(RC).

Tim Adye         26-Oct-2011
  o Fix 3D RooUnfoldResponse::Fill.
  o When inverting a matrix (eg. for chi^2), show condition, determinant, and
    check inverse.
  o RooUnfold::Print() shows a nicer summary of the object.
  o RooUnfoldTest3D: fix fake test, change test defaults, plot 2D projections.
  o RooUnfoldTest doeff=0 addfakes=1 turn off/on all inefficiencies and fakes.
  o RooUnfoldTest2D+3D projections don't include under/overflow bins.

Tim Adye         25-Oct-2011
  o Add RooUnfoldTUnfold methods for retrieving the L Curve and the tau values
    if an L curve scan is done.
    This improvement was kindly provided by Pedro Cipriano.

Tim Adye         17-Oct-2011
  o Fix Bayes response matrix errors to include the effect of the dependency of
    the prior in one iteration on the response matrix in the previous iteration.
    This is analagous to the fix of 30-Jan-2011 to the measurement errors.
    See the updated doc/bayes_errors.pdf.
  o Simplify equation 3 in doc/bayes_errors.pdf.
  o If RooUnfold::IncludeSystematics(2) is used, compute just the systematic

RooUnfold/src/RooUnfoldSvd.cxx
//=====================================================================-*-C++-*-
// File and Version Information:
//      $Id: RooUnfoldSvd.cxx 345 2013-07-29 23:38:49Z T.J.Adye $
//
// Description:
//      SVD unfolding. Just an interface to TSVDUnfold.
//
// Author: Tim Adye <T.J.Adye@rl.ac.uk>
//
//==============================================================================

//____________________________________________________________
/* BEGIN_HTML
<p>Links to TSVDUnfold class which unfolds using Singular Value Decomposition (SVD).</p>
<p>Regularisation parameter defines the level at which values are deemed to be due to statistical fluctuations and are cut out. (Default= number of bins/2)
<p>Returns errors as a full matrix of covariances
<p>Can only handle 1 dimensional distributions
<p>Can account for both smearing and biasing
END_HTML */

/////////////////////////////////////////////////////////////

#include "RooUnfoldSvd.h"

#include <iostream>
#include <iomanip>

#include "TClass.h"
#include "TNamed.h"
#include "TH1.h"
#include "TH2.h"
#include "TVectorD.h"
#include "TMatrixD.h"
#if defined(HAVE_TSVDUNFOLD) || ROOT_VERSION_CODE < ROOT_VERSION(5,34,99)
#include "TSVDUnfold_local.h"  /* Use local copy of TSVDUnfold.h */
#else
#include "TSVDUnfold.h"
#if ROOT_VERSION_CODE < ROOT_VERSION(5,34,0)
#define TSVDUNFOLD_LEAK 1
#endif
#endif

#include "RooUnfoldResponse.h"

using std::cout;
using std::cerr;
using std::endl;

ClassImp (RooUnfoldSvd);

RooUnfoldSvd::RooUnfoldSvd (const RooUnfoldSvd& rhs)
  : RooUnfold (rhs)
{
  // Copy constructor.
  Init();
  CopyData (rhs);
}

RooUnfoldSvd::RooUnfoldSvd (const RooUnfoldResponse* res, const TH1* meas, Int_t kreg,
                            const char* name, const char* title)
  : RooUnfold (res, meas, name, title), _kreg(kreg ? kreg : res->GetNbinsTruth()/2)
{
  // Constructor with response matrix object and measured unfolding input histogram.
  // The regularisation parameter is kreg.
  Init();
}

RooUnfoldSvd::RooUnfoldSvd (const RooUnfoldResponse* res, const TH1* meas, Int_t kreg, Int_t ntoyssvd,
                            const char* name, const char* title)
  : RooUnfold (res, meas, name, title), _kreg(kreg ? kreg : res->GetNbinsTruth()/2)
{
  // Constructor with old ntoyssvd argument. No longer required.
  Init();
  _NToys = ntoyssvd;
}

RooUnfoldSvd*
RooUnfoldSvd::Clone (const char* newname) const
{
  RooUnfoldSvd* unfold= new RooUnfoldSvd(*this);
  if (newname && strlen(newname)) unfold->SetName(newname);
  return unfold;
}

void
RooUnfoldSvd::Reset()
{
  Destroy();
  Init();
  RooUnfold::Reset();
}

void
RooUnfoldSvd::Destroy()
{
  delete _svd;
  delete _meas1d;
#ifdef TSVDUNFOLD_LEAK
  delete _meascov;
#endif
  delete _train1d;
  delete _truth1d;
  delete _reshist;
}

void
RooUnfoldSvd::Init()
{
  _svd= 0;
  _meas1d= _train1d= _truth1d= 0;
  _reshist= _meascov= 0;
  GetSettings();
}

void
RooUnfoldSvd::Assign (const RooUnfoldSvd& rhs)
{
  RooUnfold::Assign (rhs);
  CopyData (rhs);
}

void
RooUnfoldSvd::CopyData (const RooUnfoldSvd& rhs)
{
  _kreg= rhs._kreg;
}

TSVDUnfold*
RooUnfoldSvd::Impl()
{
  return _svd;
}

void
RooUnfoldSvd::Unfold()
{
  if (_res->GetDimensionTruth() != 1 || _res->GetDimensionMeasured() != 1) {
    cerr << "RooUnfoldSvd may not work very well for multi-dimensional distributions" << endl;
  }
  if (_kreg < 0) {
    cerr << "RooUnfoldSvd invalid kreg: " << _kreg << endl;
    return;
  }

  _nb= _nm > _nt ? _nm : _nt;

  if (_kreg > _nb) {
    cerr << "RooUnfoldSvd invalid kreg=" << _kreg << " with " << _nb << " bins" << endl;
    return;
  }

  Bool_t oldstat= TH1::AddDirectoryStatus();
  TH1::AddDirectory (kFALSE);
  _meas1d=  HistNoOverflow (_meas,             _overflow);
  _train1d= HistNoOverflow (_res->Hmeasured(), _overflow);
  _truth1d= HistNoOverflow (_res->Htruth(),    _overflow);
  _reshist= _res->HresponseNoOverflow();
  Resize (_meas1d,  _nb);
  Resize (_train1d, _nb);
  Resize (_truth1d, _nb);
  Resize (_reshist, _nb, _nb);

  // Subtract fakes from measured distribution
  if (_res->FakeEntries()) {
    TVectorD fakes= _res->Vfakes();
    Double_t fac= _res->Vmeasured().Sum();
    if (fac!=0.0) fac=  Vmeasured().Sum() / fac;
    if (_verbose>=1) cout << "Subtract " << fac*fakes.Sum() << " fakes from measured distribution" << endl;
    for (Int_t i= 1; i<=_nm; i++)
      _meas1d->SetBinContent (i, _meas1d->GetBinContent(i)-(fac*fakes[i-1]));
  }

  _meascov= new TH2D ("meascov", "meascov", _nb, 0.0, 1.0, _nb, 0.0, 1.0);
  const TMatrixD& cov= GetMeasuredCov();
  for (Int_t i= 0; i<_nm; i++)
    for (Int_t j= 0; j<_nm; j++)
      _meascov->SetBinContent (i+1, j+1, cov(i,j));

  if (_verbose>=1) cout << "SVD init " << _reshist->GetNbinsX() << " x " << _reshist->GetNbinsY()
                        << " bins, kreg=" << _kreg << endl;
  _svd= new TSVDUnfold (_meas1d, _meascov, _train1d, _truth1d, _reshist);

  TH1D* rechist= _svd->Unfold (_kreg);

  _rec.ResizeTo (_nt);
  for (Int_t i= 0; i<_nt; i++) {
    _rec[i]= rechist->GetBinContent(i+1);
  }

  if (_verbose>=2) {
    PrintTable (cout, _truth1d, _train1d, 0, _meas1d, rechist, _nb, _nb, kFALSE, kErrors);
    TMatrixD* resmat= RooUnfoldResponse::H2M (_reshist, _nb, _nb);
    RooUnfoldResponse::PrintMatrix(*resmat,"TSVDUnfold response matrix");
    delete resmat;
  }

  delete rechist;
  TH1::AddDirectory (oldstat);

  _unfolded= true;
  _haveCov=  false;
}

void
RooUnfoldSvd::GetCov()
{
  if (!_svd) return;
  Bool_t oldstat= TH1::AddDirectoryStatus();
  TH1::AddDirectory (kFALSE);

  TH2D *unfoldedCov= 0, *adetCov= 0;
  //Get the covariance matrix for statistical uncertainties on the measured distribution
  if (_dosys!=2) unfoldedCov= _svd->GetXtau();
  //Get the covariance matrix for statistical uncertainties on the response matrix
  if (_dosys) {
    TH2D* uncmat= 0;
    if (_res->Hresponse()->GetSumw2N()) {  // include weights
      Bool_t oldstat= TH1::AddDirectoryStatus();
      TH1::AddDirectory (kFALSE);
      uncmat= dynamic_cast<TH2D*>(_reshist->Clone("reserrors"));
      TH1::AddDirectory (oldstat);
      uncmat->Reset();
      for (Int_t i=0, n=_reshist->GetSize(); i<n; i++)
        uncmat->SetBinContent (i, _reshist->GetBinError (i));
    }
    adetCov= _svd->GetAdetCovMatrix (_NToys, 1, uncmat);
    delete uncmat;
  }

  _cov.ResizeTo (_nt, _nt);
  for (Int_t i= 0; i<_nt; i++) {
    for (Int_t j= 0; j<_nt; j++) {
      Double_t v  = 0;
      if (unfoldedCov) v  = unfoldedCov->GetBinContent(i+1,j+1);
      if (adetCov)     v += adetCov    ->GetBinContent(i+1,j+1);
      _cov(i,j)= v;
    }
  }

  delete adetCov;
#ifdef TSVDUNFOLD_LEAK
  delete unfoldedCov;
#endif
  TH1::AddDirectory (oldstat);

  _haveCov= true;
}

void RooUnfoldSvd::GetWgt()
{
  // Get weight matrix
  if (_dosys) RooUnfold::GetWgt();   // can't add sys errors to weight, so calculate weight from covariance
  if (!_svd) return;
  Bool_t oldstat= TH1::AddDirectoryStatus();
  TH1::AddDirectory (kFALSE);

  //Get the covariance matrix for statistical uncertainties on the measured distribution
  TH2D* unfoldedWgt= _svd->GetXinv();

  _wgt.ResizeTo (_nt, _nt);
  for (Int_t i= 0; i<_nt; i++) {
    for (Int_t j= 0; j<_nt; j++) {
      _wgt(i,j)= unfoldedWgt->GetBinContent(i+1,j+1);
    }
  }

#ifdef TSVDUNFOLD_LEAK
  delete unfoldedWgt;
#endif
  TH1::AddDirectory (oldstat);

  _haveWgt= true;
}

void RooUnfoldSvd::GetSettings(){
    _minparm=0;
    _maxparm= _meas ? _meas->GetNbinsX() : 0;
    _stepsizeparm=1;
    _defaultparm=_maxparm/2;
}

void RooUnfoldSvd::Streamer (TBuffer &R__b)
{
  // Stream an object of class RooUnfoldSvd.
  if (R__b.IsReading()) {
    // Don't add our histograms to the currect directory.
    // We own them and we don't want them to disappear when the file is closed.
    Bool_t oldstat= TH1::AddDirectoryStatus();
    TH1::AddDirectory (kFALSE);
    RooUnfoldSvd::Class()->ReadBuffer  (R__b, this);
    TH1::AddDirectory (oldstat);
  } else {
    RooUnfoldSvd::Class()->WriteBuffer (R__b, this);
  }
}
